src/Entity/Auth/UserProvider.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+namespace App\Entity\Auth;
+
+use Cocur\Slugify\SlugifyInterface;
+use Doctrine\ORM\EntityManagerInterface;
+use HWI\Bundle\OAuthBundle\OAuth\Response\UserResponseInterface;
+use HWI\Bundle\OAuthBundle\Security\Core\User\OAuthAwareUserProviderInterface;
+use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+class UserProvider implements UserProviderInterface, OAuthAwareUserProviderInterface
```

Comment:
`final`?

Reply:
I can understand `final` here (since I won't extend it anywhere).

But do you have like a generic way to choose when to define `final` class ?

Reply:
When you have an interface (contract) that expose / define your public methods you can final the class. Same for your tests.

You can read more here: https://ocramius.github.io/blog/when-to-declare-classes-final/